### ISSUE TITLE
Undocumented requirement of @pragma('vm:entry-point') for state restoration

### ISSUE THREAD

**Issue by knaeckeKami** [flutter/flutter#122434]
I was running into this issue when implementing state restoration in a flutter app.
It seems that when pushing restorable routes, the builder functions must be annotated with
`@pragma('vm:entry-point')`, otherwise state restoration does not work in release mode.

It does make sense that these functions must be annotated like this, but unless developers have at least intermediate experience in Flutter, this is not obvious to them and I did not find any mention of the requirement of these pragmas in the [docs](https://api.flutter.dev/flutter/widgets/Navigator/restorablePush.html). Not even the example has this pragma. 

## Steps to Reproduce

1. check out https://github.com/knaeckeKami/restore_root_test
2. use an android device and enable "**don't keep activities**" in the developer settings so that the app gets killed immediately when it is not in the foreground
3. run the app in **release** mode (debug mode won't trigger this bug)
4. tap on either the 'Second Page static no  pragma' or 'Second Page top level no pragma' button (the other two buttons do the same, but are on functions that are annotated with the `@pragma('vm:entry-point')` pragma - once as static function, once as top level function - and these do work as expected) 
5. but the app into background and foreground again

**Expected results:** 

The second page is restored when the app restarts.

**Actual results:** 

The second page is not restored, and the following error is printed in the console:

```
I/flutter (13112): Null check operator used on a null value
I/flutter (13112): #0      new _AnonymousRestorationInformation.fromSerializableData (package:flutter/src/widgets/navigator.dart:5390)
I/flutter (13112): #1      new _RestorationInformation.fromSerializableData (package:flutter/src/widgets/navigator.dart:5305)
I/flutter (13112): #2      _HistoryProperty.restoreEntriesForPage (package:flutter/src/widgets/navigator.dart:5552)
I/flutter (13112): #3      NavigatorState.restoreState (package:flutter/src/widgets/navigator.dart:3334)
I/flutter (13112): #4      RestorationMixin._doRestore (package:flutter/src/widgets/restoration.dart:917)
I/flutter (13112): #5      RestorationMixin.didChangeDependencies (package:flutter/src/widgets/restoration.dart:903)
I/flutter (13112): #6      NavigatorState.didChangeDependencies (package:flutter/src/widgets/navigator.dart:3405)
```

(this is on master, on stable, a different message is printed, but with the same results) 

<details>
<summary>Code sample</summary>

main.dart:
```dart
import 'package:flutter/material.dart';
import 'package:restore_root_test/second_page.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      restorationScopeId: 'app',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          ElevatedButton(
            onPressed: () {
              Navigator.of(context)
                  .restorablePush(SecondPage.secondPageRouteBuilderWithPragma);
            },
            child: const Text('Second Page static pragma'),
          ),
          ElevatedButton(
            onPressed: () {
              Navigator.of(context)
                  .restorablePush(SecondPage.secondPageRouteBuilderNoPragma);
            },
            child: const Text('Second Page static no pragma'),
          ),
          ElevatedButton(
            onPressed: () {
              Navigator.of(context)
                  .restorablePush(secondPageRouteBuilderWithPragma);
            },
            child: const Text('Second Page top level with pragma'),
          ),
          ElevatedButton(
            onPressed: () {
              Navigator.of(context)
                  .restorablePush(secondPageRouteBuilderNoPragma);
            },
            child: const Text('Second Page top level no pragma'),
          ),
        ],
      ),
    ));
  }
}


```

second_page.dart (put into a different library):

```dart
import 'package:flutter/material.dart';

Route secondPageRouteBuilderNoPragma(BuildContext context, Object? args) {
  return MaterialPageRoute<void>(
    builder: (context) => const SecondPage(),
  );
}

@pragma('vm:entry-point')
Route secondPageRouteBuilderWithPragma(BuildContext context, Object? args) {
  return MaterialPageRoute<void>(
    builder: (context) => const SecondPage(),
  );
}

class SecondPage extends StatelessWidget {
  const SecondPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Placeholder();
  }

  static Route secondPageRouteBuilderNoPragma(
      BuildContext context, Object? args) {
    return MaterialPageRoute<void>(
      builder: (context) => const SecondPage(),
    );
  }

  @pragma('vm:entry-point')
  static Route secondPageRouteBuilderWithPragma(
      BuildContext context, Object? args) {
    return MaterialPageRoute<void>(
      builder: (context) => const SecondPage(),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
Launching lib/main.dart on sdk gphone arm64 in release mode...
Running Gradle task 'assembleRelease'...
✓  Built build/app/outputs/flutter-apk/app-release.apk (6.6MB).
Installing build/app/outputs/flutter-apk/app-release.apk...
I/flutter (13112): Null check operator used on a null value
I/flutter (13112): #0      new _AnonymousRestorationInformation.fromSerializableData (package:flutter/src/widgets/navigator.dart:5390)
I/flutter (13112): #1      new _RestorationInformation.fromSerializableData (package:flutter/src/widgets/navigator.dart:5305)
I/flutter (13112): #2      _HistoryProperty.restoreEntriesForPage (package:flutter/src/widgets/navigator.dart:5552)
I/flutter (13112): #3      NavigatorState.restoreState (package:flutter/src/widgets/navigator.dart:3334)
I/flutter (13112): #4      RestorationMixin._doRestore (package:flutter/src/widgets/restoration.dart:917)
I/flutter (13112): #5      RestorationMixin.didChangeDependencies (package:flutter/src/widgets/restoration.dart:903)
I/flutter (13112): #6      NavigatorState.didChangeDependencies (package:flutter/src/widgets/navigator.dart:3405)
I/flutter (13112): #7      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5119)
I/flutter (13112): #8      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #9      Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #10     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #11     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #12     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #13     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #14     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #15     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #16     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #17     SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6377)
I/flutter (13112): #18     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #19     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #20     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #21     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5133)
I/flutter (13112): #22     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #23     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #24     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5124)
I/flutter (13112): #25     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #26     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #27     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #28     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #29     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #30     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #31     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #32     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #33     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #34     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #35     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #36     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #37     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #38     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #39     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #40     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #41     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #42     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #43     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #44     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #45     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #46     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #47     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #48     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #49     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #50     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #51     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #52     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #53     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #54     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #55     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #56     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #57     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #58     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #59     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #60     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #61     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #62     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #63     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #64     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #65     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #66     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #67     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #68     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #69     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #70     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #71     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5133)
I/flutter (13112): #72     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #73     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #74     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5124)
I/flutter (13112): #75     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #76     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #77     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #78     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #79     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #80     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #81     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #82     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #83     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #84     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #85     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #86     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #87     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #88     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #89     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #90     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #91     StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:5133)
I/flutter (13112): #92     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)
I/flutter (13112): #93     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4950)
I/flutter (13112): #94     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:5124)
I/flutter (13112): #95     ComponentElement.mount (package:flutter/src/widgets/framework.dart:4944)
I/flutter (13112): #96     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3953)
I/flutter (13112): #97     Element.updateChild (package:flutter/src/widgets/framework.dart:3682)
I/flutter (13112): #98     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4993)
I/flutter (13112): #99     Element.rebuild (package:flutter/src/widgets/framework.dart:4690)

```

```
flutter analyze
Resolving dependencies...
  async 2.10.0 (2.11.0 available)
> characters 1.3.0 (was 1.2.1)
> collection 1.17.1 (was 1.17.0)
> js 0.6.7 (was 0.6.5)
> matcher 0.12.14 (was 0.12.13)
> meta 1.9.0 (was 1.8.0)
> path 1.8.3 (was 1.8.2)
> test_api 0.4.18 (was 0.4.16)
Changed 7 dependencies!
Analyzing restore_root_test...
No issues found! (ran in 5.5s)
```

```
flutter doctor --verbose
[✓] Flutter (Channel master, 3.9.0-1.0.pre.156, on macOS 13.2.1 22D68 darwin-arm64, locale en-GB)
    • Flutter version 3.9.0-1.0.pre.156 on channel master at /Users/kami/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3b6589924a (34 minutes ago), 2023-03-10 13:52:26 -0800
    • Engine revision e9ca7b2c45
    • Dart version 3.0.0 (build 3.0.0-313.0.dev)
    • DevTools version 2.22.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/kami/Library/Android/sdk
    • Platform android-33, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] IntelliJ IDEA Community Edition (version 2021.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.75.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 13.2.1 22D68 darwin-arm64
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 111.0.5563.64

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by mraleph** [comment]
I think this is a duplicate of https://github.com/flutter/flutter/issues/118608

**Comment by knaeckeKami** [comment]
Yes, actually this was already fixed in #118738 . Maybe this would be a candidate for a CP though.
